### PR TITLE
*: remove Stash trait in favor of a single impl

### PR DESCRIPTION
**Pull request by mjibson**
We only use the Memory impl for a few tests that can now safely depend on cockroach existing.

We used the Cache impl for our production usages. That impl was added to be a quick hack to speed up operations where we didn't know what was going on. Since then we've greatly improved the underlying postgres perf and added tracing so we have complete understanding of which statements are executed. Furthermore, the Cache impl is only saving a small amount of time by caching sinces and uppers, which are one of many things we do.

Upcoming we expect to add a txn API for the stash to further reduce the total number of SQL statements. This would be unimplementable by the cache (thus not providing its caching speedup), and thus only result in more typing for us.

We can instead simplify stash callers because we now only need the single implementation.

Benchmark for adapter/catalog:

```
transact                time:   [88.400 ms 89.724 ms 91.044 ms]
                        change: [-1.4268% +4.3003% +9.9439%] (p = 0.14 > 0.05)
                        No change in performance detected.
```

So a tiny bit slower but not significant; about what we expect.

### Motivation

   * This PR refactors existing code.

### Checklist

- [ ] This PR has adequate test coverage / QA involvement has been duly considered.
- [ ] This PR evolves [an existing `$T ⇔ Proto$T` mapping](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/command-and-response-binary-encoding.md) (possibly in a backwards-incompatible way) and therefore is tagged with a `T-proto` label.
- [ ] If this PR will require changes to cloud orchestration, there is a
  companion cloud PR to account for those changes that is tagged with
  the release-blocker label ([example](https://github.com/MaterializeInc/cloud/pull/5021)).
  
- [ ] This PR includes the following [user-facing behavior changes](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/guide-changes.md#what-changes-require-a-release-note):

  - n/a

### CODE REVIEW

**Comment by mjibson** [comment]
Wanting to get this in before any txn api additions because it's so big and would like to avoid merge conflicts.

File: src/adapter/src/lib.rs [hunk 1]
```diff
@@ -74,6 +74,10 @@
 #![warn(clippy::from_over_into)]
 // END LINT CONFIG
 #![cfg_attr(nightly_doc_features, feature(doc_cfg))]
+// Without this, cargo clippy complains with:
+//     overflow evaluating the requirement `&str: std::marker::Send`
+// in implement_peek_plan's return of a Result<crate::ExecuteResponse, AdapterError>.
+#![recursion_limit = "256"]
```
Comment (mjibson) [inline src/adapter/src/lib.rs:80]: This is worrisome to me because I don't understand it.

**Review by mjibson — COMMENTED** [review]
Most of the changes are mechanical.

File: src/adapter/src/lib.rs (hunk 1 above)
Comment (jkosh44) [reply src/adapter/src/lib.rs:80]: me neither

**Review by jkosh44 — APPROVED** [review]
LGTM. Is there anyway for us to confirm that there are no serious performance regressions?

**Comment by mjibson** [comment]
Added a benchmark to the PR description.

File: src/adapter/src/lib.rs (hunk 1 above)
Comment (mjibson) [reply src/adapter/src/lib.rs:80]: This seems to be causing cargo check to take about 25s just for the adapter crate:

```
~/materialize@work (rm-stash)$ cargo check
    Checking mz-adapter v0.0.0 (/home/mjibson/materialize/src/adapter)
    Checking mz-pgwire v0.0.0 (/home/mjibson/materialize/src/pgwire)
    Checking mz-testdrive v0.0.0 (/home/mjibson/materialize/src/testdrive)
    Checking mz-stash-debug v0.28.0-dev (/home/mjibson/materialize/src/stash-debug)
    Checking mz-environmentd v0.40.0-dev (/home/mjibson/materialize/src/environmentd)
    Checking mz-sqllogictest v0.0.1 (/home/mjibson/materialize/src/sqllogictest)
    Finished dev [unoptimized + debuginfo] target(s) in 33.10s
```

(they all went quite fast after adapter)

**Comment by mjibson** [comment]
New commit added some debug catalog stuff which was needed to support parallel test runs.

**Pull request merged**